### PR TITLE
list algorithms + benchmark

### PR DESCRIPTION
**Pull request by karunrao97**

### CODE REVIEW

File: course3/homework-list/list.h [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+
+namespace Karun
+{
+
+template<typename T>
+class List
+{
+public:
+	typedef T value_type;
```
Comment (karunrao97) [inline course3/homework-list/list.h]: This was required in order for std::front_inserter to work. Is there a better way?

File: course3/homework-list/test.cc [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+	}
+	
+	// distance
+	l = {1, 1, 1, 1, 2, 2, 2, 3, 3, 4};
+	assert(std::distance(l.begin(), l.end()) == 10);
+	
+	// count and count_if
+	assert(std::count(l.begin(), l.end(), 2) == 3);
+	assert(std::count_if(l.begin(), l.end(), [](int n){ return n >= 2; }) == 6);
+	//assert(boost::count(l, 2) == 3);
```
Comment (karunrao97) [inline course3/homework-list/test.cc]: The code doesn't compile for Boost range based algorithms. Any hints on what's required?

File: course3/homework-list/benchmark.cc [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+template<template<typename...> class ListType>
+void ListBenchmark(benchmark::State& state)
+{
+	ListType<int> l;
+	for (int i=0; i < state.range_x(); ++i)
+		l.push_front(i);
+}
+
+//BENCHMARK_TEMPLATE(ListBenchmark, Karun::List)->Range(1<<0, 1<<10);
```
Comment (karunrao97) [inline course3/homework-list/benchmark.cc]: BENCHMARK_TEMPLATE and BENCHMARK_MAIN don't compile. I get an "Undefined symbols for architecture x86_64" error. I'm running the code on OS X with clang++.

File: course3/homework-list/list.h (hunk 1 above)
Comment (maciekgajewski) [inline course3/homework-list/list.h]: No. This is the way. Actually, I personally prefer the "using" version:

```
using value_type = T;
```

Some code needs these to infer stored type (and other types) from the container type.

File: course3/homework-list/test.cc (hunk 2 above)
Comment (maciekgajewski) [inline course3/homework-list/test.cc]: The main requirements are 'begin' and 'end' and well-behaving iterators. I can't say what's wrong without seeing the error message.

File: course3/homework-list/benchmark.cc (hunk 3 above)
Comment (maciekgajewski) [inline course3/homework-list/benchmark.cc]: I can't see any inking with the benchmark library in your CMakeLists.txt

File: course3/homework-list/test.cc (hunk 2 above)
Comment (karunrao97) [inline course3/homework-list/test.cc]: Error message attached.
[error.txt](https://github.com/maciekgajewski/c-excercises/files/335649/error.txt)

File: course3/homework-list/benchmark.cc (hunk 3 above)
Comment (karunrao97) [inline course3/homework-list/benchmark.cc]: I added the target_link_libraries (see updated CMakeLists.txt), but now get the following linker error.
[error.txt](https://github.com/maciekgajewski/c-excercises/files/335664/error.txt)

Comment (maciekgajewski) [inline course3/homework-list/benchmark.cc]: The error is pretty clear: `library not found for -lbenchmark`
Do you have the library installed in a place where the linker can find it? On Linux, I'd be looking for a file named `libbenchmark.a` or `libbenchmark.so.X.Y` in one of the standard system paths. I don't know what the convention is on your OS.

File: course3/homework-list/test.cc (hunk 2 above)
Comment (maciekgajewski) [inline course3/homework-list/test.cc]: I must commend the clang/boost combo for producing really readable error messages:

```
error: cannot increment value of type 'Karun::List<int>::iterator'
                (void)i++;
                      ~^
```

Your iterator class doesn't provide _postfix_ increment operator. I don't know why boost wants to use postfix (`i++`) instead if prefix (`++i`) version. but if they require it, then let's provide it.

http://en.cppreference.com/w/cpp/language/operators

Comment (karunrao97) [inline course3/homework-list/test.cc]: Thanks, fixed. Boost algos working now :)

File: course3/homework-list/benchmark.cc (hunk 3 above)
Comment (karunrao97) [inline course3/homework-list/benchmark.cc]: I have `libbenchmark.a` in `/usr/local/lib`, which is in my `$PATH`. Is there an `lbenchmark` file also expected somewhere?

Comment (karunrao97) [inline course3/homework-list/benchmark.cc]: Fixed by adding `/usr/local/lib` to `$LIBRARY_PATH` instead.

**Comment by karunrao97** [comment]
Benchmarks "work" now (see output here: [bench.txt](https://github.com/maciekgajewski/c-excercises/files/337796/bench.txt))

Unfortunately, as you can see in the output, my list implementation segfaults for large lists. Turns out it segfaults as soon as we get to exactly 100k elements. Any idea why this magic number?

**Comment by maciekgajewski** [comment]
Is your benchmark build in debug or release mode?

**Comment by maciekgajewski** [comment]
Do you have a core file from the crash? Have you tried running it in debugger to see hwere exactly the crash happens?

**Comment by maciekgajewski** [comment]
I suspect that the recursive destruction of elements runs out of stack. There are 2 solutions: compile it in release mode and hope that compiler will replace the recursion with iteration, or manually write recursion-free destructor.

**Comment by karunrao97** [comment]
I tried compiling in release mode, but it still failed. I then implemented an iterative destructor, and it works fine! Thanks :)

**Comment by karunrao97** [comment]
Updated benchmark results (for `push_front`, `copy`, `move` and `transform`) here: [bench.txt](https://github.com/maciekgajewski/c-excercises/files/338098/bench.txt)

In general, performance is similar to `std::forward_list`, slightly better than `std::list`, and far worse than `std::deque`.

**Comment by maciekgajewski** [comment]
Great job!

**Pull request closed without merging**